JKA-838 講師側お知らせ削除API ロジック作成（小林）

Celadon67 opened the pull request:
## issue
- https://gut-familie.atlassian.net/browse/JKA-838
## 概要
- 講師側 お知らせ 一括（選択）削除API作成　子課題
  ロジック作成
## 動作確認手順
- DELETEリクエストで　http://localhost:8080/api/v1/instructor/notification
  を実行すると以下の返答が返ってきた

```
{
    "result": true
}
```

- ただ今現在はNotificationを全て削除したため、以下のエラー文が表示されている

```
{
    "result": false,
    "message": "Failed to delete notifications. invalid notification_id."
}
```

## 考慮してほしいこと
- 今回は特にありません。
## 確認してほしいこと
- 今回は特にありません。

Celadon67 committed:
JKA-838 講師側お知らせ削除API 空の配列を返すルーターとコントローラの作成

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成

github-actions[bot] committed:
Automated code format

yukihiroLaravel commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,22 @@ public function update(NotificationUpdateRequest $request): JsonResponse
     }
+
+    /**
+    * お知らせ全削除
```
もともと全削除でしたが
分かりやすく
名前を今は「選択」削除にしましたので
選択削除に変更お願いします！

Celadon67 replied:
変更しました。

yukihiroLaravel commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,22 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+    {
+        $notificationIds = $request->input('notifications', []);
+
+        $user = Auth::guard('instructor')->user();
```
今のままですと
notificationを作った本人でない場合、削除できないですが
エラーが出ずに、削除できないだけで終わってしまい
何が原因で失敗したのかわからないので、
if文を用いた認証を書き、
さらにその中でresultがfalseになる
エラーを返すパターンも作成しましょう。

Celadon67 replied:
①削除対象のお知らせがログインユーザーのものかどうかを確認
②削除対象のお知らせが存在しない、またはログインユーザーと一致しない場合はエラーを返す
③削除対象のお知らせが存在し、かつログインユーザーと一致する場合は削除を実行
の考え方でコード変更しました。

yukihiroLaravel commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,22 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+                ->where('instructor_id', $user->id)
+                ->delete();
+
+        return response()->json(['result' => true]);
```
他のところを見てもらうと下記のような書き方なので
変更しましょう！
```
            return response()->json([
                'result' => true,
            ]);
```

Celadon67 replied:
修正しました。

yukihiroLaravel commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,22 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+        $notificationIds = $request->input('notifications', []);
+
+        $user = Auth::guard('instructor')->user();
+        Notification::whereIn('id', $notificationIds)
+                ->where('instructor_id', $user->id)
+                ->delete();
```
上記に伴いこちらの記述方法も変わってくるかと思います

Celadon67 replied:
上記に伴い変更させました。

yukihiroLaravel reviewed (changes requested):
いくつか指摘させていただきました！
ご確認お願いします！

Celadon67 committed:
JKA-838 講師側お知らせ削除API 空のルーター・コントローラー　選択削除に名前変更

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 エラー原因が分かるようにコード修正

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 コンフリクト解消

yukihiroLaravel commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,36 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+        if ($notifications->isEmpty()) {
+            return response()->json([
+                'result' => false,
+                'message' => 'Failed to delete notifications. Either the notifications do not exist or you are not authorized to delete them.',
```
~ or ~
のようなエラーの出し方だと
エラーの原因特定がやりにくくなるので、ここでは
エラーメッセージにdo not exit などの文言は不要になると思います。

多分、
invalid notification_id.
などの文言で十分かな？と思います。

Celadon67 replied:
修正しました。

yukihiroLaravel commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,36 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+
+        Notification::whereIn('id', $notificationIds)
+                    ->where('instructor_id', $user->id)
+                    ->delete();
```
１１６行目と同じ内容で検索してしまっているので無駄はなくしましょう！

Celadon67 replied:
削除しました。

yukihiroLaravel commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,36 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+
+        $user = Auth::guard('instructor')->user();
+
+        $notifications = Notification::whereIn('id', $notificationIds)
+                        ->where('instructor_id', $user->id)
+                        ->get();
```
他の部分と統一性を持たせるために
if文で １つ１つのnotificationのinstructor_idと、$user->idの照合を掛けるのをまず
行った方がいいと思いました。

Celadon67 replied:
修正いたしました。

yukihiroLaravel reviewed (changes requested):
いくつかコメントしました！

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 if文で一つのnotificationのinstructor_idと、->idの照合を行う 他修正込み

kou-tech commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,36 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+
+        $user = Auth::guard('instructor')->user();
+
+        $notifications = Notification::whereIn('id', $notificationIds)->get();
+
+        $notifications = $notifications->filter(function ($notification) use ($user) {
+            return $notification->instructor_id === $user->id;
+        });
```
同じ変数名の格納は困惑を招くので避けた方が良いです。
そもそも同じ変数名だが処理の前後で変わるのであれば、バグの一因になることが多々あります。

Celadon67 replied:
修正しました。

kou-tech commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,36 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+            ], 403);
+        }
+
+        $notifications->each->delete();
```
こちらはeachの記述の必須でしたか？

Celadon67 replied:
削除しました。

Celadon67 committed:
JKA-838 講師側お知らせ削除API 空の配列を返すルーターとコントローラの作成 メソッドを一括削除に合わせて変更

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 変数名の変更+他

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 前タスクのマージに伴うコンフリクト解消

kou-tech commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,38 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+
+        $user = Auth::guard('instructor')->user();
+
+        $allNotifications = Notification::whereIn('id', $notificationIds)->get();
+
+        $filteredNotifications = $allNotifications->filter(function ($notification) use ($user) {
+            return $notification->instructor_id === $user->id;
+        });
+
+        if ($filteredNotifications->isEmpty()) {
+            return response()->json([
+                'result' => false,
+                'message' => 'Failed to delete notifications. invalid notification_id.',
+            ], 403);
+        }
```
処理は間違っていないのですが、他の開発者がコードを見た時に理解するという面では大変そうな雰囲気がありますよね。
好みも含みますが、次のようなコレクションメソッドを使用しつつ、コメントで補足するとどうでしょうか？

```php
        if ($notifications->contains(function (Notification $notification) use ($user) {
            return $notification->instructor_id !== $user->id;
        })) {
            // 認証者と一致しないお知らせが含まれている場合はエラー
            return response()->json([
                'result' => false,
                'message' => 'Forbidden, not allowed to delete notifications.',
            ], 403);
        }
```

Celadon67 replied:
コメントで補足しました。

kou-tech commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +100,38 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+            ], 403);
+        }
+
+        Notification::whereIn('id', $notificationIds)
+        ->where('instructor_id', $user->id)
+        ->delete();
```
お知らせは既に取得しているので、再度DBにアクセスする必要性はないかと思います！

Celadon67 replied:
修正しました。

kou-tech replied:
@Celadon67 
再度DBにアクセスするは、修正されていないのでご確認お願いします！
分からないことがあれば、また質問してください。

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 データベースへのアクセスを削減・コメントで補足追記

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 データベースへのアクセスを削減②

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 データベースへのアクセスを削減③

github-actions[bot] committed:
Automated code format

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 データベースへのアクセスを削減③ インテンド修正

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 データベースへのアクセスを削減③ インテンド修正後のコンフリクト解消

github-actions[bot] committed:
Automated code format

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 データベースへのアクセスを削減 指定された通知IDのみを使用して直接削除

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 データベースへのアクセスを削減・コメントで補足追記　ここに戻す

github-actions[bot] committed:
Automated code format

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 通知IDを使用して再度データベースから通知を取得せず、直接指定された通知IDを使用して削除

Celadon67 committed:
Merge branch 'feature/seiji/JKA-838/instructor_announcements_delete_api_logic' of https://github.com/yukihiroLaravel/juko_laravel into feature/seiji/JKA-838/instructor_announcements_delete_api_logic

github-actions[bot] committed:
Automated code format

Commit:
Fix : 次の点を修正。
- userというテーブルはないことから、変数に明示的にinstructorを命名するように修正。
-  インストラクター、講師といった似たような表現を避けるべく講師に統一
- 認可処理を関数型アプローチで実装するように対応。
- 削除を取得したお知らせのリストから、eachで削除するように変更。
- 明示的なコメントはノイズとなるため、削除。
- 悪意のある認可エラーであることが推測される場合には簡素なエラーメッセージに修正。

github-actions[bot] committed:
Automated code format

Commit:
Fix : bulkDeleteに対応。

kou-tech commented:
- `viewed_once_notifications`のレコードを削除
- トランザクションを導入

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 viewed_once_notificationsのレコードを削除+トランザクションを導入

github-actions[bot] committed:
Automated code format

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 viewed_once_notificationsのレコードを削除+トランザクションを導入 認証者を講師に変更

Celadon67 committed:
Merge branch 'feature/seiji/JKA-838/instructor_announcements_delete_api_logic' of https://github.com/yukihiroLaravel/juko_laravel into feature/seiji/JKA-838/instructor_announcements_delete_api_logic

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 削除を取得したお知らせのリストから、eachで削除するように変更

kou-tech commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +102,60 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+        try {
+            // お知らせを一括削除
+            $notifications->each(function ($notification) {
+                DB::table('viewed_once_notifications')->where('notification_id', $notification->id)->delete();
```
現状システム全体として、一貫して`Eloquent`を介してデータベースへの操作を行っているので、こちらも同様に`Eloquent`でご対応いただけますか？

複雑なSQL文などを理由に、クエリビルダで書く時もあるので間違いではないのですが、現状はEloquentで書くようにお願いいたします！モデルがない場合、モデルの作成からお願いいたします🙇‍♂️

Celadon67 replied:
モデルがないため加えました。それに伴いコントローラー、Notificationモデルも変更いたしました。

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 Eloquentを介してデータベースへの操作のための修正

github-actions[bot] committed:
Automated code format

kou-tech commented on app/Http/Controllers/Api/Instructor/NotificationController.php:
```diff
@@ -99,4 +103,60 @@ public function update(NotificationUpdateRequest $request): JsonResponse
+        DB::beginTransaction();
+
+        try {
+            // お知らせを一括削除
+            $notifications->each(function ($notification) {
+                $notification->viewedOnceNotifications()->delete();
+                $notification->delete();
+            });
```
これ以前共有したQiitaの記事の通り、次の記述でデータベースのレイヤーで一括削除が実行できます。
https://qiita.com/atkzw3/items/c4e4a5c0f3a866b9b95a 

```php
            Notification::whereIn('id', $notificationIds)->delete();
```

そのため、次の手順で削除するのがパフォーマンスとして、利点がありそうなので、修正いただけますか？
1. viewedOnceNotificationsテーブルのレコードを一括削除
2. notificationsテーブルのレコードを一括削除

Celadon67 replied:
viewedOnceNotificationsのSeederを作成し、一括削除のコードに修正いたしました。

kou-tech reviewed (changes requested):
viewedOnceNotificationsのSeederを一つで構いませんので、作成していただくようにお願いします！

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 viewedOnceNotificationsのSeederを作成+データベースのレイヤーで一括削除

Celadon67 committed:
Merge branch 'feature/seiji/JKA-838/instructor_announcements_delete_api_logic' of https://github.com/yukihiroLaravel/juko_laravel into feature/seiji/JKA-838/instructor_announcements_delete_api_logic

kou-tech commented on database/seeds/ViewedOnceNotificationSeeder.php:
Seederを再実行すると次のエラーが出力されるので、ご確認いただけますか？

```txt
 Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails
```

kou-tech replied:
不整合データは動作確認のノイズとなりやすいので、避けたいです！
お知らせの既読状態を管理しているのは、`notifications`テーブルの`type`が`once`の時だけです。

なので、`notification_id=2`と`student_id=2`のレコードを一件だけインサートするシーダーでお願いいたします！

Celadon67 replied:
修正しました。

kou-tech commented on database/migrations/2023_09_03_013745_create_viewed_once_notifications_table.php:
```diff
@@ -15,10 +15,9 @@ public function up()
-            $table->dateTime('updated_at');
+            $table->unsignedBigInteger('notification_id');
+            $table->unsignedBigInteger('student_id');
+            $table->timestamps();
```
こちら修正前後で列の型が変わってしまっているのは、意図的ですか？
timestampだと次の懸念があり、他の実装で意図せずtimestampを使ってしまっていますが、なるべくdatetime型でお願いしたいです！
https://zenn.dev/kenghaya/articles/89bc9a6e9e8381

Celadon67 replied:
修正しました。

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 Seederエラー解決・timestamp変更

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 列の型戻し

kou-tech commented on composer.json:
```diff
@@ -35,7 +35,8 @@
     "autoload": {
         "psr-4": {
-            "App\\": "app/"
+            "App\\": "app/",
+            "Database\\Seeders\\": "database/seeders/"
```
おそらく新規に作成したSeederが動作しなかった対応だと思いますが、composerのキャッシュの問題だと思われるので、こちらは修正を取り消していただき、次のコマンドでご自身のcomposerを更新しましょう！

```shell
$ composer dump-autoload
```

Celadon67 replied:
修正を取り消して変更がなくなりました。

Celadon67 committed:
JKA-838 講師側お知らせ削除API ロジック作成 composerを更新

Commit:
Fix : 例外時エラーを出力し、必要なインポートの記述を追加。

github-actions[bot] committed:
Automated code format
